### DISCUSSION TITLE
How to test code paths that require pytest dependencies *NOT* be installed?

### DISCUSSION THREAD

**Discussion by posita** [pytest-dev/pytest#9725] [Q&A]
I have a package that will opportunistically take advantage of `typing-extensions` when running in Python 3.7 if available, but will operate at reduced functionality without it. I would like to be able to run my entire battery of tests to ensure they all pass both with and without `typing-extensions`.

_**However**_, as far as I can tell, `pytest` sweeps in `typing-extensions` for Python 3.7. (More specifically, it [requires `importlib-metadata`](https://github.com/pytest-dev/pytest/blob/e466a87bb082ea5f7377f84038322e9bc8fba789/setup.cfg#L52), which [requires `typing-extensions`](https://github.com/python/importlib_metadata/blob/14cce75299645467adcd17352cb07caada32c444/setup.cfg#L21).)

How does one go about ~testing code paths that are only executed~ _running an entire suite of tests_ when `typing-extensions` is _**unavailable**_ in that environment?

Consider: I have a core component `A` in my project `proj`. Pretty much everything in `proj` depends on `A`. Under the covers I have two different implementations of `A`: `AwithB` if dependency `B` is found, and a reduced-functionality stand-in `AwithoutB` if it is absent. (`B` is an optional dependency of `proj`.) How do I go about running all the tests I have for `proj` if `B` is always swept in by `pytest`?

I feel like I can't be the first person to encounter or think about this.

_UPDATE: Could this be done using something akin to a context manager that temporarily sets `sys.modules['typing_extensions']` to `None` (e.g., something [like this](https://github.com/beartype/beartype/pull/107#issuecomment-1055642600)). Maybe I can configure that via [`conftest.py`](https://docs.pytest.org/en/6.2.x/writing_plugins.html#hookwrapper-executing-around-other-hooks) or a [hook wrapper](https://docs.pytest.org/en/6.2.x/writing_plugins.html#hookwrapper-executing-around-other-hooks)?_

**Comment by RonnyPfannschmidt** [comment]
More context is needed to give a good answer, is this a tool, or just about type specs

**Reply by posita** [reply]
This is an expression of an unmet need. It's an observation about the limitations of `pytest` (i.e., inclusion of its own dependencies in the testing environment that are inseparable from those available to the code being tested) and how that fights with my particular requirement that one such dependence very much not be present in the testing environment.

This is a question about practices for addressing that shortcoming. I have no opinion about the form of a solution. I'm open.

**Comment by 93578237** [comment]
Check this: https://youtu.be/PXu3KCMT3l4

**Reply by posita** [reply]
That's actually a pretty cool trick, but that how does one mock the absence of a dependency for the entire test suite?

**Comment by posita** [comment]
I suppose one answer is to refactor the implementation to enable something like dependency injection for the with- and without-optional dependency cases. Then I suppose you could use some environment-based configuration (e.g., an environment variable or some other initialization machinery) to override injecting the set of dependencies you're trying to test. But this isn't always practical, especially when inheriting a large codebase.
